modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+                throw new IgniteCheckedException("Unable to copy file [snapshot=" + opCtx.snpName +
+                    ", grp=" + FilePageStoreManager.cacheGroupName(cacheDir) + ']', e);
+            }
+        }
+    }
+
+    /**
+     * Rollback changes made by process in specified cache group.
+     *
+     * @param opCtx Snapshot restore operation context.
+     */
+    private void rollback(@Nullable SnapshotRestoreContext opCtx) {
+        if (opCtx == null || F.isEmpty(opCtx.dirs))
+            return;
+
+        if (log.isInfoEnabled())
+            log.info("Performing local rollback routine for restored cache groups [requestID=" + opCtx.reqId + ']');
+
+        try {
+            for (File cacheDir : opCtx.dirs) {
+                if (!cacheDir.exists())
+                    continue;
+
+                if (log.isInfoEnabled())
+                    log.info("Cleaning up directory " + cacheDir);
+
+                U.delete(cacheDir);
+            }
+        }
+        catch (Exception e) {
+            log.error("Failed to perform rollback [requestID=" + opCtx.reqId + ", snapshot=" + opCtx.snpName + ']', e);
+        }
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @param res Results.
+     * @param errs Errors.
+     */
+    private void finishPrepare(UUID reqId, Map<UUID, ArrayList<StoredCacheData>> res, Map<UUID, Exception> errs) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (ctx.clientNode())
+            return;
+
+        Exception failure = checkFailure(errs, opCtx0, res.keySet());
+
+        if (failure == null) {
+            assert opCtx0 != null : ctx.localNodeId();
+
+            Map<Integer, StoredCacheData> globalCfgs = new HashMap<>();
+
+            for (List<StoredCacheData> storedCfgs : res.values()) {
+                if (storedCfgs == null)
+                    continue;
+
+                for (StoredCacheData cacheData : storedCfgs)
+                    globalCfgs.put(CU.cacheId(cacheData.config().getName()), cacheData);
+            }
+
+            opCtx0.cfgs = globalCfgs;
+
+            if (U.isLocalNodeCoordinator(ctx.discovery()))
+                cacheStartProc.start(reqId, reqId);
+
+            return;
+        }
+
+        if (opCtx0 == null)
+            finishProcess(failure);
+        else // Remove files asynchronously.
+            ctx.cache().context().snapshotMgr().snapshotExecutorService().execute(() -> {
+                rollback(opCtx0);
+
+                finishProcess(failure);
+            });
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @return Result future.
+     */
+    private IgniteInternalFuture<Boolean> cacheStart(UUID reqId) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 == null)
+            return new GridFinishedFuture<>();
+
+        if (!reqId.equals(opCtx0.reqId)) {
+            return new GridFinishedFuture<>(
+                new IgniteCheckedException("Unknown snapshot restore operation was rejected."));
```

Comment:
The `restore` procedure state is incorrect if this condition occurred. I don't think another concurrent restore operation can be executed here, so it's better to fix the error messsage.

Reply:
done, this check has been removed